### PR TITLE
Add support for setting arbitrary NameResolver attributes in channel builder

### PR DESCRIPTION
**Pull request by elandau**
It is currently not possible to pass in any custom parameters to a NameResolver.  This change adds the ability to specify NameResolver parameters on the base ManagedChannelBuilder.  

Note that the current API delegates the responsibility for setting such parameters to the concrete ManagedChannelBuilder (such as NettyChannelBuilder).  This is not ideal as it assumes a NameResolver's configuration it dependent on the Channel builder type.  For backwards compatibility I kept the existing AbstractManagedChannelImplBuilder#getNameResolverParams() method and added a new ManagedChannelBuilder#nameResolverParams() method for setting NameResolver parameters. Parameters from both mechanisms are merged when build() is called with parameters passed to nameResolverParams() taking precedence.

### CODE REVIEW

**Comment by jhspaybar** [comment]
I think the approach of a ChannelBuilder implementation overriding that method is fine, it needs the opportunity to inject parameters that could affect the servers discovered.  I wonder though if we wouldn't be better off following a URI style format though to pass attributes from the layer you're at here.  What type of data do you see getting injected through attributes at this level?

**Comment by elandau** [comment]
I'm actually in the process of adding support for URI format style of passing attributes :)   It might actually be sufficient for us to more forward without needing this PR.  I still find their API strange in that on one hand they expose the ability to pass in configuration via the Attributes passed to the NameResolver.Factory but don't provide a mechanism to actually provide the configuration and even worse tie it to the Channel implementation.

**Comment by jhspaybar** [comment]
I might open an issue on gRPC expressing the concerns and thoughts, or even just throw this PR itself up against the gRPC repo to gather comments.  We might be missing part of the story and/or expectations they have.  It'd be great to gather that information if possible.

File: core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java [hunk 1]
```diff
@@ -94,6 +93,9 @@
   @Nullable
   private CompressorRegistry compressorRegistry;
 
+  @Nullable
```
Comment (tbak) [inline core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java]: From what I see it is not nullable?

**Pull request closed without merging**